docs/devel/api_changes.md

```diff
@@ -0,0 +1,237 @@
+	Param  string `json:"param"`
+}
+```
+
+The onus is on you to define a sane default value for `Width` such that rule #1
+above is true - API calls that used to work must continue to work.
+
+For your next change you want to allow multiple `Param` values.  You can not
+simply change `Param string` to `Params []string` without reviing the whole API
+- that fails rules #1 and #2.  You can instead do something like:
+
+```go
+// Still API v6, but kind of clumsy.
+type Frobber struct {
+	Height int           `json:"height"`
+	Width  int           `json:"width"`
+	Param  string        `json:"param"`  // the first param
+	ExtraParams []string `json:"params"` // additional params
+}
+```
+
+Now you can satisfy the rules: API calls that provide the old style `Param`
+will still work, while servers that don't understand `ExtraParams` can ignore
+it.  This is somewhat unsatisfying as an API, but it is strictly compatible.
+
+Part of the reason for versioning APIs and for using an explicit internal API
+is to handle growth like this.  The internal API can be implemented as:
+
+```go
+// Internal API, soon to be v7beta1.
+type Frobber struct {
+	Height int           `json:"height"`
+	Width  int           `json:"width"`
+	Params []string      `json:"params"`
+}
+```
+
+The code that converts to/from versioned APIs can decode this into the somewhat
+uglier (but compatible!) structures.  Eventually, a new API version, let's call
+it v7beta1, will be forked and it can use the internal structure.
+
+We've seen how to satisfy rules #1 and #2.  Rule #3 means that you can not
+extend one versioned API without also extending the others.  For example, an
+API call might POST an object in API v7beta1 format, which uses the cleaner
+`Params` field, but the API server might store that object in trusty old v6
+form (since v7beta1 is "beta").  When the user reads the object back in the
+v7beta1 API it would be unacceptable to have lost all but `Params[0]`.  This
+means that, even though it is ugly, a compatible change must be made to the v6
+API.
+
+## Changing versioned APIs
+
+For most changes, you will probably find it easiest to change the versioned
+APIs first.  This forces you to think about how to make your change in a
+compatible way.  Rather than doing each step in every version, it's usually
+easier to do each versioned API one at a time, or to do all of one version
+before starting "all the rest".
+
+### Edit types.go
+
+The struct definitions for each API are in `pkg/api/<version>/types.go`.  Edit
+those files to reflect the change you want to make.
+
+### Edit defaults.go
+
+If your change includes new fields for which you will need default values, you
+need to add cases to `pkg/api/<version>/defaults.go`.  Of course, since you
+have added code, you have to add a test: `pkg/api/<version>/defaults_test.go`.
+
+Don't forget to run the tests!
+
+### Edit conversion.go
+
+Given that you have not yet changed the internal API, this might feel premature
```

Comment:
We should just call it the "internal representation" or "internal API representation" everywhere.